src/downloader/sentry_client_reactor.rs

```diff
@@ -191,56 +191,120 @@ impl Drop for SentryClientReactor {
     }
 }
 
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
+enum EventLoopStreamId {
+    Send,
+    Receive,
+    Stop,
+}
+
+#[derive(Debug)]
+enum EventLoopStreamResult {
+    Send(u32),
+    Receive(MessageFromPeer),
+    Stop(()),
+}
+
+type EventLoopStream = Pin<Box<dyn Stream<Item = anyhow::Result<EventLoopStreamResult>> + Send>>;
+
 impl SentryClientReactorEventLoop {
     async fn run(&mut self) -> anyhow::Result<()> {
+        let mut sentry = self.sentry.take().unwrap();
+
         // subscribe to incoming messages
-        let mut in_stream = self.sentry.receive_messages(&[]).await?;
-
-        loop {
-            tokio::select! {
-                Some(command) = self.send_message_receiver.recv() => {
-                    let send_result = self.sentry.send_message(command.message, command.peer_filter).await;
-                    match send_result {
-                        Ok(sent_peers_count) => {
-                            debug!("SentryClientReactor.EventLoop sent message to {:?} peers", sent_peers_count);
+        let mut stream = sentry.receive_messages(&[]).await?;
+        let receive_messages_senders = Arc::clone(&self.receive_messages_senders);
+        let receive_stream = async_stream::stream! {
+            while let Some(message_result) = stream.next().await {
+                yield message_result;
+            }
+
+            debug!("SentryClientReactor.EventLoop receive_messages stream ended");
+            // drop shared senders so that existing receive_messages streams end
+            receive_messages_senders.write().clear();
```

Comment:
Stuff like this should be in `Drop` impl of some dummy struct because Stream API provides no guarantees as to execution to the end.

```
struct SomeDummyDropper(Arc<RwLock<HashMap<EthMessageId, broadcast::Sender<Message>>>>)

impl Drop for SomeDummyDropper {
    fn drop(&mut self) {
            // drop shared senders so that existing receive_messages streams end
            receive_messages_senders.write().clear();
    }
}

let receive_stream = async_stream::stream! {
    let d = SomeDummyDropper(receive_messages_senders);

    ...
}
```